system/lib/fetch/wasmfs.cpp

```diff
@@ -0,0 +1,55 @@
+// // Copyright 2021 The Emscripten Authors.  All rights reserved.
```

Comment:
The issue that @curiousdannii brings up about using C++ in core libraries is worth thinking about.

The issues is that up until now one could build a C program and link agianst only C libraries and avoid the C++ runtime completely.  (i.e. use `emcc` rather then `em++`.. as of today we still make C++ libraries available unless `-s STRICT` mode is also used but I hope to change that).

Using C++ language features should be fine but if we use C++ library features in our filesystem implementation then it makes it much hard to build a pure C application. 

We do have some core libraries such as embind that are writting in C++ but those are not used to implement low level C APIs and its generally accepted that embind only works for C++ programs.

Its maybe not a huge issue, but if we do use C++ (and C++ library features in particular) that does means that users of WASMFS would need to make sure they link in libc++ and libc++abi.

Reply:
Interesting point. Does that mean that even using `new` would be an issue, as that method is in libc++abi I think? If so that is concerning.

I do think it would be a big productivity win to use as modern a language as we can here.

Reply:
Yes, I agree it nice to be able to implement system libraries in C++... and I think we should probably allow it here, but traditionally the C library implementation is implement in C .. for this reason.. otherwise it makes all program into C++ programs.

Really we are implementing part of the "OS" in C++ here but they happen to get linked into the user program.  

I guess we can just have `-s WASMFS` imply `-s DEFAULT_TO_CXX`.

Reply:
> I guess we can just have -s WASMFS imply -s DEFAULT_TO_CXX.

Yes. But note that the goal is to eventually enable WASMFS by default, and then even remove the option so it is the only thing we support, which would mean that any program using a nontrivial file-related syscall would end up a C++ program. So we need to make a decision here I guess.

Reply:
I guess it depends how much benefit we see from using C++ here.  Perhaps its not the end of the world if all filesystem-using programs end up depending on C++.

Reply:
Maybe we can bundle C++ runtime dependencies into the wasmfs build somehow so they don't need to be separately linked in?

Reply:
`DEFAULT_TO_CXX` should be sufficient to guarantee they are linked it.

Reply:
Could LTO versions of libc++ etc be included in the Docker build? Prior to this you'd only have to wait for it to be built when your own code was linking against libc++, but now it will be all C projects that use the FS too.

Reply:
Yes, the Docker build should contains all flavors of libc++ i think.  IIUC its a bug that it doesn't.